src/content/datachannel/messaging/main.js

```diff
@@ -0,0 +1,180 @@
+
+      let lc = this._localChannel = this._localConnection
+        .createDataChannel('messaging-channel', dataChannelParams);
+      lc.binaryType = 'arraybuffer';
+      lc.addEventListener('open', e => this._localChannelOpen(e));
+      lc.addEventListener('close', e => this._channelClosed(e));
+      lc.addEventListener('message', e => this._onLocalMessageReceived(e));
+
+      rpc.addEventListener('datachannel', e => this._onRemoteDataChannel(e));
+
+      const localOffer = await lpc.createOffer();
+      console.log(`Got local offer ${JSON.stringify(localOffer)}`);
+      lpc.setLocalDescription(localOffer);
+      rpc.setRemoteDescription(localOffer);
+      const remoteOffer = await rpc.createAnswer();
+      console.log(`Got remote answer ${JSON.stringify(remoteOffer)}`);
+      rpc.setLocalDescription(remoteOffer);
+      lpc.setRemoteDescription(remoteOffer);
+    } catch (e) {
+      console.log(e);
+    }
+  }
+
+  async _onIceCandidate(e, connection) {
+    try {
+      console.log(`onIceCandidate: ${JSON.stringify(e)}`);
+      // eslint-disable-next-line no-unused-vars
+      const result = await connection.addIceCandidate(e.candidate);
+      console.log('addIceCandidate successful!');
+    } catch (e) {
+      console.log(`error on addIceCandidate: ${JSON.stringify(e)}`);
+    }
+  }
+
+  _localChannelOpen(event) {
+    console.log(`Local channel open: ${JSON.stringify(event)}`);
+    this.connected = true;
+    // this.shadowRoot.querySelector('#sendLocal').disabled = false;
+    // this.shadowRoot.querySelector('#sendRemote').disabled = false;
+    // this.shadowRoot.querySelector('#disconnectButton').disabled = false;
+  }
+
+  _channelClosed(event) {
+    console.log(`Channel closed: ${JSON.stringify(event)}`);
+    this.connected = false;
+    // this.shadowRoot.querySelector('#sendLocal').disabled = true;
+    // this.shadowRoot.querySelector('#sendRemote').disabled = true;
+    // this.shadowRoot.querySelector('#connectButton').disabled = false;
+  }
+
+  _onLocalMessageReceived(event) {
+    console.log(`Remote message received by local: ${event.data}`);
+    this.shadowRoot.querySelector('#localIncoming').value += event.data + '\n';
+  }
+
+  _onRemoteDataChannel(event) {
+    console.log(`onRemoteDataChannel: ${JSON.stringify(event)}`);
+    this._remoteChannel = event.channel;
+    this._remoteChannel.binaryType = 'arraybuffer';
+    this._remoteChannel.addEventListener('message', e => this._onRemoteMessageReceived(e));
+    this._remoteChannel.addEventListener('close', e => this._channelClosed(e));
+  }
+
+  _onRemoteMessageReceived(event) {
+    console.log(`Local message received by remote: ${event.data}`);
+    this.shadowRoot.querySelector('#remoteIncoming').value += event.data + '\n';
+  }
+
+  static get properties() {
+    return {
+      connected: {
+        type: Boolean,
+        notify: true,
+        reflectToAttribute: true
```

Comment:
Why reflecting to attribute?

Reply:
It means that your MessagingSample (ie <message-sample> can have a connected attribute like:)

`<message-sample connected>`

Reply:
Also do you need `notify`? isnt it not only for callbacks? like observing when it changes from the outside